transaction/std/src/memo_builder/gift_code_memo_builder.rs

```diff
@@ -0,0 +1,181 @@
+/// | 0x0002          | Gift Code Sender Memo                             |
+/// | 0x0201          | Gift Code Funding Memo                            |
+/// | 0x0202          | Gift Code Cancellation Memo                       |
+#[derive(Clone, Debug)]
+pub enum GiftCodeMemoType {
+    /// Gift Code Sender Memo 0x0002
+    Sender,
+    /// Gift Code Funding Memo 0x0201
+    Funding,
+    /// Gift Code Cancellation Memo 0x0202
+    Cancellation,
+}
+
+/// This memo builder writes gift code memos to the change address which can be
+/// used by apps that offer gift codes as feature to track the status of gift
+/// codes as well as certain contextual information about them. This builder
+/// writes 3 possible gift code memo types. Only one memo which can be specified
+/// at the creation of the memo builder. It is worth noting that all memos
+/// written by this memo builder will go to the reserved change address and
+/// does not write any destination memos. For normal recoverable transaction
+/// history (RTH), the RTH memo builder should be used in conjuction with this
+/// memo builder.
+#[derive(Clone, Debug)]
+pub struct GiftCodeMemoBuilder {
+    // The gift code memo type we'll be writing
+    gift_code_memo_type: GiftCodeMemoType,
+    // The utf-8 formatted note that appears on the change address when a gift code is
+    // or cancelled
+    note: &'static str,
+    // TxOut Public Key of the gift code if it it's a funding memo
+    gift_code_tx_out_public_key: Option<RistrettoPublic>,
+    // Global TxOut index of the gift code if it's a cancellation memo
+    gift_code_tx_out_global_index: Option<u64>,
+    // Whether or not to enable gift code change memos
+    gift_code_change_memo_enabled: bool,
+    // Whether we've already written the change memo
+    wrote_change_memo: bool,
+}
+
+impl GiftCodeMemoBuilder {
+    /// Create a new gift code memo builder specifying the memo type
+    pub fn new(gift_code_memo_type: GiftCodeMemoType) -> Self {
+        Self {
+            gift_code_memo_type,
+            note: "",
+            gift_code_tx_out_public_key: None,
+            gift_code_tx_out_global_index: None,
+            gift_code_change_memo_enabled: true,
+            wrote_change_memo: false,
+        }
+    }
+    /// A note in the form of null terminated utf-8 bytes can be applied to
+    /// gift code funding or sender memos. These notes end up in the change
+    /// outputs and indicate what the memo was for. Set this memo here. This
+    /// method will enforce the 60 and 64 byte length for Funding Memos and
+    /// Sender Memos respectively. If the memo type is a cancellation memo
+    /// and a note is set, any value set here will be ignored as the
+    /// cancellation memo does not include a note.
+    pub fn set_note(&mut self, note: &'static str) -> Result<(), MemoError> {
+        let max_note_length = if let GiftCodeMemoType::Funding = self.gift_code_memo_type {
+            60
```

Comment:
This seems like duplication of magic numbers.  Suggest using  `GiftCodeFundingMemo::NOTE_DATA_LEN`, and creating a companion `GiftCodeSenderMemo::NOTE_DATA_LEN`.  
Not really fond of creating a new const on the sender memo just for the builder, but this repeating of these values here seems like a bad code smell.
i.e. what happens if the memo size gets bumped to 128 and this gets changed but the memos themselves get missed.

Reply:
This is a great point!

Reply:
I think it'll be fine to have the MEMO_LEN const for the sender memo as as it works right now, all bytes are reserved for utf-8 so 128 bytes would be reserved. But yeah excellent catch on the constant usage being best practice.

Reply:
The issue I saw with the MEMO_LEN is that the `GiftCodeFundingMemo` also has a `MEMO_LEN` which is the full 64 bytes, it may be tempting for someone to change that if they see `NOTE_DATA_LEN` from one and `MEMO_LEN` from another

Reply:
I've refactored them into separate transaction builders to clear up the ambiguity